src/box/memtx_tx.c

```diff
@@ -2996,13 +2961,87 @@ memtx_tx_delete_gap(struct gap_item_base *item)
 	memtx_tx_mempool_free(item->txn, pool, item);
 }
 
+/**
+ * Abort all concurrent readers of the old schema of the space.
+ */
 void
-memtx_tx_invalidate_space(struct space *space, struct txn *active_txn)
+memtx_tx_abort_space_schema_readers(struct space *space, struct txn *ddl_owner)
 {
+	/* Abort all story and gap-with-sucessor readers. */
+	struct memtx_story *story;
+	rlist_foreach_entry(story, &space->memtx_stories, in_space_stories) {
+		struct tx_read_tracker *tracker, *tmp;
+		rlist_foreach_entry_safe(tracker, &story->reader_list,
+					 in_reader_list, tmp) {
+			if (tracker->reader != ddl_owner)
+				memtx_tx_abort_with_conflict(tracker->reader);
```

Comment:
Shouldn't we skip prepared transactions?

Reply:
Oh, thanks!
Only active transactions can have trackers - when transaction is prepared or rolled back, all its trackers are deleted. So it makes sense only for iteration over `all_txs`. Fixed:
```diff
diff --git a/src/box/memtx_tx.c b/src/box/memtx_tx.c
index 6ce2cfbde2..98411acbea 100644
--- a/src/box/memtx_tx.c
+++ b/src/box/memtx_tx.c
@@ -773,6 +773,7 @@ memtx_tx_adjust_position_in_read_view_list(struct txn *txn)
 static void
 memtx_tx_abort_with_conflict(struct txn *victim)
 {
+	assert(victim->psn == 0);
 	if (victim->status == TXN_ABORTED)
 		return;
 	if (victim->status == TXN_IN_READ_VIEW)
@@ -2963,8 +2964,10 @@ memtx_tx_delete_gap(struct gap_item_base *item)
 
 /**
  * Abort all concurrent readers of the old schema of the space.
+ * Note that only active transaction can have trackers since when a transaction
+ * is rolled back or prepared all its trackers are deleted.
  */
-void
+static void
 memtx_tx_abort_space_schema_readers(struct space *space, struct txn *ddl_owner)
 {
 	/* Abort all story and gap-with-sucessor readers. */
@@ -3008,6 +3011,9 @@ memtx_tx_abort_space_schema_readers(struct space *space, struct txn *ddl_owner)
 	 */
 	struct txn *txn;
 	rlist_foreach_entry(txn, &txm.all_txs, in_all_txs) {
+		if (txn->status != TXN_INPROGRESS &&
+		    txn->status != TXN_IN_READ_VIEW)
+			continue;
 		if (txn == ddl_owner)
 			continue;
 		struct txn_stmt *stmt;
```

The new assertion covers the problem.

Reply:
Would be nice to have a test using `ERRINJ_WAL_DELAY` that would ensure this. There's a test for this in Vinyl:

https://github.com/tarantool/tarantool/blob/017700dccdb04edd8470af620c69f3e7c572895b/test/vinyl/errinj_ddl.test.lua#L156-L176

Reply:
Yeah, sounds right. Done.